pennylane/interfaces/jacobian_products.py

```diff
@@ -231,22 +245,31 @@ def execute_and_compute_jvp(self, tapes: Batch, tangents: Tuple[Tuple[TensorLike
     def compute_vjp(self, tapes: Batch, dy: Tuple[Tuple[TensorLike]]):
         if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
             logger.debug("compute_vjp called with (%s, %s)", tapes, dy)
+
+        if self._cache_full_jacobian:
+            jacs = self.compute_jacobian(tapes)
+            return _compute_vjps(jacs, dy, tapes)
+
         vjp_tapes, processing_fn = qml.gradients.batch_vjp(
             tapes, dy, self._gradient_transform, gradient_kwargs=self._gradient_kwargs
         )
 
-        vjp_results = self._inner_execute(vjp_tapes)
+        vjp_results = self._inner_execute(tuple(vjp_tapes))
         return tuple(processing_fn(vjp_results))
 
     def compute_jacobian(self, tapes: Batch):
         if logger.isEnabledFor(logging.DEBUG):  # pragma: no cover
             logger.debug("compute_jacobian called with %s", tapes)
+        if tapes in self._cache:
+            return self._cache[tapes]
         partial_gradient_fn = partial(self._gradient_transform, **self._gradient_kwargs)
         jac_tapes, batch_post_processing = qml.transforms.map_batch_transform(
             partial_gradient_fn, tapes
         )
         results = self._inner_execute(jac_tapes)
-        return tuple(batch_post_processing(results))
+        jacs = tuple(batch_post_processing(results))
+        self._cache[tapes] = jacs
```

Comment:
I'm not super familiar with LRUCache perf/functionality, but I'm surprised to see the tapes themselves as keys. is this how we cache tape results elsewhere? I assumed we'd use `tuple(tape.hash for t in tapes)` or something like that.

Reply:
This is essentially "cheap caching". `tape.hash` can add in substantial overheads, so instead I'm instead just caching based on the in-built hash, since in this case it will be the same batch getting submitted.

Reply:
This is for the use case when `interface == "autograd" and not cache`.

To mimic the closure-based caching that was there before, I'm purely caching based on instance identity, since the same instances will be submitted each time inside `vjp.grad_fn.